### ISSUE TITLE
Code highlighting and many other featuress completely broken with specific project

### ISSUE THREAD

**Issue by terrarier2111** [intellij-rust/intellij-rust#9718]
## Environment

* **IntelliJ Rust plugin version: 0.4.182.4996-222**
* **Rust toolchain version: Nightly**
* **Operating system: Fedora Linux**
* ![Bildschirmfoto vom 2022-11-13 01-45-45](https://user-images.githubusercontent.com/58695553/201500270-f372689e-002c-4325-a376-3427750db3cf.png)

## Problem description
When opening (and even interacting with) a certain file inside a certain project, the highlighting is completely broken and the idea becomes basically unusable (it provides no help finding function/type usage, it has no highlighting, refactoring doesn't work properly 
![Bildschirmfoto vom 2022-11-13 01-49-01](https://user-images.githubusercontent.com/58695553/201500402-0b6b081a-0323-4824-bf80-b57a73a5f0b6.png)
etc).

## Steps to reproduce
Go to the file `swap_arc_tls_optimistic.rs` in the project located at: https://github.com/terrarier2111/SwapArc

**Comment by neonaot** [comment]
There is an exception during type inference:
```
java.lang.IllegalStateException: Can not run nested type inference
	at org.rust.lang.core.types.infer.TypeInferenceKt.inferTypesIn(TypeInference.kt:41)
```

<details>

``` 

java.lang.IllegalStateException: Can not run nested type inference
	at org.rust.lang.core.types.infer.TypeInferenceKt.inferTypesIn(TypeInference.kt:41)
	at org.rust.lang.core.types.ExtensionsKt._get_inference_$lambda$0(Extensions.kt:112)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:231)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:41)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:231)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:234)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.types.ExtensionsKt.getInference(Extensions.kt:111)
	at org.rust.lang.core.types.ExtensionsKt.getInference(Extensions.kt:126)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.rawMultiResolveUsingInferenceCache(RsPathReferenceImpl.kt:117)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.rawMultiResolve(RsPathReferenceImpl.kt:113)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.resolve(RsPathReferenceImpl.kt:98)
	at org.rust.lang.utils.evaluation.PathExprResolver$Companion$default$1.invoke(Utils.kt:14)
	at org.rust.lang.utils.evaluation.PathExprResolver$Companion$default$1.invoke(Utils.kt:14)
	at org.rust.lang.utils.evaluation.PathExprResolver.invoke(Utils.kt)
	at org.rust.lang.utils.evaluation.ConstExprBuilder.buildInner(ConstExprBuilder.kt:58)
	at org.rust.lang.utils.evaluation.IntegerConstExprBuilder.buildInner(ConstExprBuilder.kt:106)
	at org.rust.lang.utils.evaluation.ConstExprBuilder.build(ConstExprBuilder.kt:49)
	at org.rust.lang.utils.evaluation.ConstExprBuilder.buildInner(ConstExprBuilder.kt:56)
	at org.rust.lang.utils.evaluation.IntegerConstExprBuilder.buildInner(ConstExprBuilder.kt:106)
	at org.rust.lang.utils.evaluation.ConstExprBuilder.build(ConstExprBuilder.kt:49)
	at org.rust.lang.utils.evaluation.ConstExprBuilder.build(ConstExprBuilder.kt:44)
	at org.rust.lang.utils.evaluation.ConstExprBuilderKt.toConstExpr(ConstExprBuilder.kt:27)
	at org.rust.lang.utils.evaluation.ConstExprEvaluatorKt.evaluate(ConstExprEvaluator.kt:27)
	at org.rust.lang.utils.evaluation.ConstExprEvaluatorKt.toConst(ConstExprEvaluator.kt:33)
	at org.rust.lang.core.types.infer.TyLowering.instantiatePathGenerics(TyLowering.kt:242)
	at org.rust.lang.core.types.infer.TyLowering.access$instantiatePathGenerics(TyLowering.kt:38)
	at org.rust.lang.core.types.infer.TyLowering$Companion.lowerPathGenerics(TyLowering.kt:287)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedResolve(RsPathReferenceImpl.kt:89)
	at org.rust.lang.core.resolve.NameResolutionKt.processQualifiedPathResolveVariants(NameResolution.kt:431)
	at org.rust.lang.core.resolve.NameResolutionKt.processPathResolveVariants(NameResolution.kt:359)
	at org.rust.lang.core.resolve.NameResolutionKt.processPathResolveVariants(NameResolution.kt:343)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt$resolvePathRaw$1.invoke(RsPathReferenceImpl.kt:288)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt$resolvePathRaw$1.invoke(RsPathReferenceImpl.kt:287)
	at org.rust.lang.core.resolve.ProcessorsKt.collectResolveVariantsAsScopeEntries(Processors.kt:189)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt.resolvePathRaw(RsPathReferenceImpl.kt:287)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferPathExprType(TypeInferenceWalker.kt:323)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:182)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType$default(TypeInferenceWalker.kt:171)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferCallExprType(TypeInferenceWalker.kt:610)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:188)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.processStatement(TypeInferenceWalker.kt:150)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:118)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType$default(TypeInferenceWalker.kt:114)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferIfExprType(TypeInferenceWalker.kt:1039)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:192)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType$default(TypeInferenceWalker.kt:171)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferTypeCoercableTo(TypeInferenceWalker.kt:223)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:122)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferTypeCoercableTo(TypeInferenceWalker.kt:112)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferFnBody(TypeInferenceWalker.kt:107)
	at org.rust.lang.core.types.infer.RsInferenceContext.infer(TypeInference.kt:217)
	at org.rust.lang.core.types.infer.TypeInferenceKt.inferTypesIn$lambda$0(TypeInference.kt:40)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:112)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:42)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:66)
	at org.rust.openapiext.UtilsKt.recursionGuard(utils.kt:94)
	at org.rust.lang.core.types.infer.TypeInferenceKt.inferTypesIn(TypeInference.kt:40)
	at org.rust.lang.core.types.ExtensionsKt._get_inference_$lambda$0(Extensions.kt:112)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:231)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:41)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:231)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:112)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:42)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:66)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:232)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.types.ExtensionsKt.getInference(Extensions.kt:111)
	at org.rust.lang.core.types.ExtensionsKt.getInference(Extensions.kt:126)
	at org.rust.lang.core.types.ExtensionsKt.getType(Extensions.kt:138)
	at org.rust.ide.hints.type.RsChainMethodTypeHintsProviderKt.getType(RsChainMethodTypeHintsProvider.kt:169)
	at org.rust.ide.hints.type.RsChainMethodTypeHintsProviderKt.access$getType(RsChainMethodTypeHintsProvider.kt:1)
	at org.rust.ide.hints.type.RsChainMethodTypeHintsProvider$getCollectorFor$1.collect(RsChainMethodTypeHintsProvider.kt:99)
	at com.intellij.codeInsight.hints.CollectorWithSettings.collectHints(InlayHintsUtils.kt:92)
	at com.intellij.codeInsight.hints.InlayHintsPass.doCollectInformation$lambda-0(InlayHintsPass.kt:47)
	at com.intellij.concurrency.ApplierCompleter.execAndForkSubTasks(ApplierCompleter.java:136)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1086)
	at com.intellij.concurrency.ApplierCompleter.lambda$wrapInReadActionAndIndicator$1(ApplierCompleter.java:92)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:591)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:666)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:622)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:590)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.concurrency.ApplierCompleter.wrapInReadActionAndIndicator(ApplierCompleter.java:104)
	at com.intellij.concurrency.ApplierCompleter.compute(ApplierCompleter.java:86)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)

```

</details>

Related to https://github.com/intellij-rust/intellij-rust/issues/9685?

**Comment by vlad20012** [comment]
Maybe fixed by #9714. @neonaot could you please check it?

**Comment by neonaot** [comment]
Yes, works fine with the latest nightly plugin:
<img width="1025" alt="image" src="https://user-images.githubusercontent.com/49211026/201655507-b11ac5d4-451f-4e2c-9a96-977a4d8b392c.png">

@terrarier2111, you can install nightly builds using [this instruction](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-quick-start.html#install-nightly).
